src/Streamly/Internal/Data/Fold/Concurrent/Channel/Type.hs

```diff
@@ -0,0 +1,330 @@
+    , svarRef :: Maybe (IORef ())
+
+    -- Stats --
+    , svarStats :: SVarStats
+
+    -- Diagnostics --
+    , svarInspectMode :: Bool
+    , svarCreator :: ThreadId
+    }
+
+{-# NOINLINE dumpSVar #-}
+dumpSVar :: Channel m a b -> IO String
+dumpSVar sv = do
+    xs <- sequence $ intersperse (return "\n")
+        [ return (dumpCreator (svarCreator sv))
+        , return "---------CURRENT STATE-----------"
+        , dumpOutputQ (outputQueue sv)
+        -- XXX print the types of events in the outputQueue, first 5
+        , dumpDoorBell (outputDoorBell sv)
+        , return "---------STATS-----------\n"
+        , dumpSVarStats (svarInspectMode sv) Nothing (svarStats sv)
+        ]
+    return $ concat xs
+
+-------------------------------------------------------------------------------
+-- Support for running folds concurrently
+-------------------------------------------------------------------------------
+
+-- $concurrentFolds
+--
+-- To run folds concurrently, we need to decouple the fold execution from the
+-- stream production. We use the SVar to do that, we have a single worker
+-- pushing the stream elements to the SVar and on the consumer side a fold
+-- driver pulls the values and folds them.
+--
+-- @
+--
+-- Fold worker <------Channel<------Fold driver
+--     |  exceptions  |
+--     --------------->
+--
+-- @
+--
+-- We need a channel for pushing exceptions from the fold worker to the fold
+-- driver. The stream may be pushed to multiple folds at the same time. For
+-- that we need one Channel per fold:
+--
+-- @
+--
+-- Fold worker <------Channel--
+--                    |        |
+-- Fold worker <------Channel------Driver
+--                    |        |
+-- Fold worker <------Channel--
+--
+-- @
+--
+-- Note: If the stream pusher terminates due to an exception, we do not
+-- actively terminate the fold. It gets cleaned up by the GC.
+
+-------------------------------------------------------------------------------
+-- Process events received by a fold worker from a fold driver
+-------------------------------------------------------------------------------
+
+sendToDriver :: Channel m a b -> ChildEvent b -> IO Int
+sendToDriver sv msg = do
+    -- In case the producer stream is blocked on pushing to the fold buffer
+    -- then wake it up so that it can check for the stop event or exception
+    -- being sent to it otherwise we will be deadlocked.
+    -- void $ tryPutMVar (pushBufferMVar sv) ()
+    sendWithDoorBell (outputQueueFromConsumer sv)
+                     (outputDoorBellFromConsumer sv) msg
+
+sendYieldToDriver :: MonadIO m => Channel m a b -> b -> m ()
+sendYieldToDriver sv res = liftIO $ do
+    void $ sendToDriver sv (ChildYield res)
+
+{-# NOINLINE sendExceptionToDriver #-}
+sendExceptionToDriver :: Channel m a b -> SomeException -> IO ()
+sendExceptionToDriver sv e = do
+    tid <- myThreadId
+    void $ sendToDriver sv (ChildStop tid (Just e))
+
+data FromSVarState m a b =
```

Comment:
Change `SVar` to `Channel`? Should be a small change and anyone who further reads this will not get confused.